core/variables.js

```diff
@@ -337,77 +447,55 @@ Blockly.Variables.createVariable = function(workspace, opt_callback, opt_type) {
  *     aborted (cancel button), or undefined if an existing variable was chosen.
  */
 Blockly.Variables.renameVariable = function(workspace, variable,
-  opt_callback) {
-  // (karishma) TODO (#1244) Modal message should change depending on what type
-  // of variable is getting renamed.
-  // This function needs to be named so it can be called recursively.
-  var promptAndCheckWithAlert = function(defaultName) {
-    var promptText =
-        Blockly.Msg.RENAME_VARIABLE_TITLE.replace('%1', variable.name);
-    Blockly.Variables.promptName(promptText, defaultName,
-        function(newName) {
-          if (newName) {
-            if (workspace.getVariable(newName, variable.type)) {
-              Blockly.alert(Blockly.Msg.VARIABLE_ALREADY_EXISTS.replace('%1',
-                  newName),
-                  function() {
-                    promptAndCheckWithAlert(newName);  // Recurse
-                  });
-            } else {
-              workspace.renameVariableById(variable.getId(), newName);
-            }
-            if (opt_callback) {
-              opt_callback(newName);
-            }
-          } else {
-            // User canceled prompt without a value.
-            if (opt_callback) {
-              opt_callback(null);
-            }
-          }
-        }, null);
-  };
-  promptAndCheckWithAlert('');
-};
+    opt_callback) {
+  // Validation and modal message/title depends on the variable type
+  var promptMsg, modalTitle, validate;
+  var varType = variable.type;
+  if (varType == Blockly.LIST_VARIABLE_TYPE) {
+    promptMsg = Blockly.Msg.RENAME_LIST_TITLE;
+    modalTitle = Blockly.Msg.RENAME_LIST_MODAL_TITLE;
+  } else if (varType == Blockly.SCALAR_VARIABLE_TYPE) {
+    promptMsg = Blockly.Msg.RENAME_VARIABLE_TITLE;
+    modalTitle = Blockly.Msg.RENAME_VARIABLE_MODAL_TITLE;
+  } else {
+    // Only lists and variables can be renamed... this is an error
+    console.warn('Encountered unexpected variable type ' + varType +
+        'when attempting to rename a variable.');
+    return;
+  }
+  var validate = Blockly.Variables.nameValidator_(varType);
```

Comment:
Line 286 is 
`  var validate = Blockly.Variables.nameValidator_.bind(null, opt_type);`

Why is this one handled differently?

Reply:
Oops! that was unintentional. Fixed.